src/sage/matrix/matrix_integer_dense.pyx

```diff
@@ -3250,6 +3281,18 @@ cdef class Matrix_integer_dense(Matrix_dense):
             R = A.to_matrix(self.new_matrix())
             if transformation:
                 U = U.to_matrix(self.new_matrix(U.nrows, U.ncols))
+
+        elif algorithm == 'pari':
+            # call pari with flag=4: kernel+image
+            # pari uses column convention: need to transpose the matrices
+            from sage.libs.pari import pari
```

Comment:
if we use `self._gp_().mattranspose()` below, then, we don't need to import pari on this line

Also we would need to use `K.sage().augment(T.sage())` or something like this.

I don't have a strong opinion on this. We may keep it as it is if it is better overall.

Reply:
I'm not sure about this one. Maybe someone else has an opinion?

Reply:
**Never** use `_gp_`, which is going through the old and slow interface ; use `__pari__` 

```
sage: m=matrix(ZZ,4,4,range(16))
sage: type(m)
<class 'sage.matrix.matrix_integer_dense.Matrix_integer_dense'>
sage: m._gp_()
[0, 1, 2, 3; 4, 5, 6, 7; 8, 9, 10, 11; 12, 13, 14, 15]
sage: type(_)
<class 'sage.interfaces.gp.GpElement'>
sage: m.__pari__()
[0, 1, 2, 3; 4, 5, 6, 7; 8, 9, 10, 11; 12, 13, 14, 15]
sage: type(_)
<class 'cypari2.gen.Gen'>

```

Reply:
Ok, thanks! Would it be better to use __pari__ to avoid the import statement?

Reply:
I don't really see a reason to avoid the import statement: It's essentially free since the library has already been loaded anyway.
```
sage: %timeit from sage.libs.pari import pari
182 ns ± 1.79 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
Shouldn't this be using `cypari2`, not `pari` ?

Reply:
> Shouldn't this be using `cypari2`, not `pari` ?

What do you mean?
```python
>>> from sage.libs.pari import pari
>>> type(pari)
<class 'cypari2.pari_instance.Pari'>
```

Reply:
oops, never mind, sorry

Reply:
Though, for the conversion, it is probably faster to go through `M.__pari__()` or even the more direct
```
cdef Gen pari_matrix = integer_matrix(self._matrix, 0)
```
(which is the one-line code of `M.__pari__()`). The second argument of `integer_matrix` allows you to transpose which might be useful here!

Reply:
To convert back from pari the most direct at the moment is to call `gen_to_sage_matrix` from `sage.libs.pari.convert_sage_matrix`. Though this function does not take into account the fact that the entries are integral and there is a bunch of irrelevant python objects built for each entry. For huge matrices it could matter.

Reply:
I will propose a commit.